src/group-by.mjs

```diff
@@ -1,39 +1,99 @@
 import { iterableCurry } from './internal/iterable'
+import { Exchange } from './internal/queues'
 
-function * groupBy (key = (k) => k, iterable) {
+function groupBy (getKey = (k) => k, iterable) {
   const iterator = iterable[Symbol.iterator]()
 
+  let itemIndex = 0
+  let iterableCounter = 0
+  let noNewIterables = false
+  const exchange = new Exchange()
+  const consumer = exchange.getConsumer()
+  let done = false
   let currentItem
-  let currentKey, previousKey
-
-  function * group () {
-    while (true) {
-      yield currentItem.value
-      currentItem = iterator.next()
-      if (currentItem.done) return
-      currentKey = key(currentItem.value)
-      if (previousKey !== currentKey) {
-        return
+
+  // fetch new item from Iterator
+  // return the item and advance the
+  // main queue consumer
+  function fetch () {
+    const newItem = iterator.next()
+    if (newItem.done) {
+      done = true
+      return
+    }
+    exchange.push({ value: newItem.value, key: getKey(newItem.value, itemIndex++) })
+    currentItem = consumer.shift()
+  }
+
+  // close the original iterator if possible
+  function returnIterator () {
+    if (noNewIterables && iterableCounter === 0) {
+      if (typeof iterator.return === 'function') iterator.return()
+    }
+  }
+
+  // generate subgroup where adjacent items have the same key
+  // it picks up new item from the buffer. Every instance has
+  // its own independent consumer
+  function * generateGroup (firstItem, cons) {
+    try {
+      iterableCounter++
+      // the function generator is ready.
+      // *1*: I use this trick to ensure that finally is called
+      yield 'ready'
+
+      yield firstItem.value
+
+      while (true) {
+        if (cons.isEmpty()) {
+          fetch()
+          if (done) return
+          continue
+        }
+        const nextItem = cons.shift()
+        if (nextItem.key !== firstItem.key) {
+          return // see *2*
+        }
+        yield nextItem.value
       }
+    } finally {
+      iterableCounter--
+      returnIterator()
     }
-  };
-
-  try {
-    currentItem = iterator.next()
-
-    while (true) {
-      if (currentItem.done) return
-      currentKey = key(currentItem.value)
-      if (previousKey !== currentKey) {
-        previousKey = currentKey
-        yield [currentKey, group()]
-      } else {
-        currentItem = iterator.next()
+  }
+
+  // generate group returning [key, subgroup]
+  // it always picks up new items from the Iterator (fetch)
+  function * generateGroups () {
+    // using an empty object as initial key:
+    // it is surely different from any possible key
+    let currentKey = {}
```

Comment:
I think you could just call this `key`. I also don't really mind if you leave it this way though. I have just begun to think that `current` is usually meaningless. If you just say `item` and `key` it will be obvious that the current item is whatever is in the variable. And you can still have `nextItem` or `previousItem` of course.

Also I think self documenting code techniques would work here too. You could make this a constant like `const UNIQUE_INITIAL_KEY = {}`.

Reply:
I can't use key, because I have many other keys around, and it would be confusing,

I used currentKey, because this is the key of the main loop that is currently in use.

I like the idea of using a constant UNIQUE_INITIAL_KEY :+1: